src/RestrictedPython/transformer.py

```diff
@@ -385,6 +385,10 @@ def transform_slice(self, slice_):
                 dims.elts.append(self.transform_slice(item))
             return dims
 
+        # Python 3.9.0a5+ for `a[b]`, `a['b']` resp. `a[1,2]`:
+        elif isinstance(slice_, (ast.Name, ast.Constant, ast.Tuple)):
```

Comment:
Investigating further: what formerly was modeled as an `ExtSlice` is now modeled as a `Tuple` containing (one or more) `Slice` items. This implies that `Tuple` should get the treatment of `ExtSlice`.

Python 3.6:
```
>>> from ast import parse, dump
>>> print (dump(parse("l[1, 2:3]", mode="eval")))
Expression(body=Subscript(value=Name(id='l', ctx=Load()), slice=ExtSlice(dims=[Index(value=Num(n=1)), Slice(lower=Num(n=2), upper=Num(n=3), step=None)]), ctx=Load()))
```

Python 3.9:
```
>>> from ast import parse, dump
>>> print (dump(parse("l[1, 2:3]", mode="eval"), indent=4))
Expression(
    body=Subscript(
        value=Name(id='l', ctx=Load()),
        slice=Tuple(
            elts=[
                Constant(value=1),
                Slice(
                    lower=Constant(value=2),
                    upper=Constant(value=3))],
            ctx=Load()),
        ctx=Load()))
```
However, almost surely, the method must change significantly (if it is still necessary at all): beside `Name`, `Constant` and `Tuple` almost any expression type can arrive as the following (3.9) example demonstrates:
```
>>> print (dump(parse("l[a+b, 2:3]", mode="eval"), indent=4))
Expression(
    body=Subscript(
        value=Name(id='l', ctx=Load()),
        slice=Tuple(
            elts=[
                BinOp(
                    left=Name(id='a', ctx=Load()),
                    op=Add(),
                    right=Name(id='b', ctx=Load())),
                Slice(
                    lower=Constant(value=2),
                    upper=Constant(value=3))],
            ctx=Load()),
        ctx=Load()))
```
This implies that it must handle all those types.

Do we still know why this transform was introduced? I have the hope that it may no longer be necessary.

If it is still necessary, we likely should change the logic: handle known cases which require actions (`Tuple`, `Slice`, `ExtSlice`, `Index`) and return everything else unchanged.

Reply:
With Python 3.9, `transform_slice` is (almost surely) no longer needed (i.e. can return its parameter unchanged) as explained below.
`transform_slice` is (only) used by `visit_Subscript`. Its purpose is to wrap subscriptions into function calls, for reading subscriptions into calls to `_getitem_`. Formerly, `Slice` was only allowed inside a `Subscript` and could not be directly used as parameter to `_getitem_`; with Python 3.9, `Slice` is a subtype of `expr` and can likely be directly used as parameter to `_getitem_`.

I suggest to check for version >= 3.9 and return in this case `slice_` unchanged and otherwise use the current code. If this approach is invalid, the tests should show this (likely with a `TypeError` when the `_getitem_` call is compiled).